patch/openurlonclick.c

```diff
@@ -1,59 +1,140 @@
+		url[--i] = TLINEURL(row_start)[col_start].u;
+		if (--col_start < 0) {
+			if (--row_start < minrow || (col_start = findeowl(row_start)) < 0)
+				break;
+		}
+	} while (i > 0 && isvalidurlchar(TLINEURL(row_start)[col_start].u));
+
+	/* early detection */
+	if (url[i] != 'h')
+		return NULL;
+
+	/* find the last character of url */
+	do {
+		x2 = col, y2 = row;
+		url_maxcol = MAX(url_maxcol, x2);
+		url[j++] = TLINEURL(row)[col].u;
+		wrapped = TLINEURL(row)[col].mode & ATTR_WRAP;
+		#if VIM_BROWSE_PATCH
+		if (++col >= buffCols || wrapped) {
+		#elif COLUMNS_PATCH
+		if (++col >= term.maxcol || wrapped) {
+		#else
+		if (++col >= term.col || wrapped) {
+		#endif // VIM_BROWSE_PATCH
+			col = 0;
+			if (++row > maxrow || !wrapped)
+				break;
+		}
+	} while (j < sizeof(url)-1 && isvalidurlchar(TLINEURL(row)[col].u));
+
+	url[j] = 0;
+
+	if (strncmp("https://", &url[i], 8) && strncmp("http://", &url[i], 7))
+		return NULL;
+
+	/* underline url (see xdrawglyphfontspecs() in x.c) */
+	if (draw) {
+		url_x1 = (y1 >= 0) ? x1 : 0;
+		url_x2 = (y2 < term.row) ? x2 : url_maxcol;
+		url_y1 = MAX(y1, 0);
+		url_y2 = MIN(y2, term.row-1);
+		url_draw = 1;
+		for (y1 = url_y1; y1 <= url_y2; y1++)
+			term.dirty[y1] = 1;
+	}
+
+	return &url[i];
+}
+
 void
 openUrlOnClick(int col, int row, char* url_opener)
 {
-   int row_start = row;
-   int col_start = col;
-   int row_end = row;
-   int col_end = col;
-
-   if (term.line[row][col].u == ' ')
-       return;
-
-   /* while previous character is not space */
-   while (term.line[row_start][col_start-1].u != ' ') {
-       if (col_start == 0)
-       {
-           // Before moving start pointer to the previous line we check if it ends with space
-           if (term.line[row_start - 1][term.col - 1].u == ' ')
-               break;
-           col_start=term.col - 1;
-           row_start--;
-       } else {
-           col_start--;
-       }
-   }
-
-   /* while next character is not space nor end of line */
-   while (term.line[row_end][col_end].u != ' ') {
-       col_end++;
-       if (col_end == term.col - 1)
-       {
-           if (term.line[row_end + 1][0].u == ' ')
-               break;
-           col_end=0;
-           row_end++;
-       }
-   }
-
-   char url[200] = "";
-   int url_index=0;
-   do {
-       url[url_index] = term.line[row_start][col_start].u;
-       url_index++;
-       col_start++;
-       if (col_start == term.col)
-       {
-           col_start = 0;
-           row_start++;
-       }
-   } while (url_index < (sizeof(url)-1) &&
-            (row_start != row_end || col_start != col_end));
-
-   if (strncmp("http", url, 4) != 0) {
-       return;
-   }
-
-   char command[strlen(url_opener)+strlen(url)+2];
-   sprintf(command, "%s %s", url_opener, url);
-   system(command);
+	char *url = detecturl(col, row, 1);
+	if (url) {
+		char command[strlen(url_opener) + strlen(url) + 5];
+		sprintf(command, "%s \"%s\"&", url_opener, url);
+		system(command);
```

Comment:
I don't think you need `system`, `posix_spawn` should be function you should be looking at. Which would also eliminate the need for `command` and `sprintf` as well since `posix_spawn` take an `argv`.